### DISCUSSION TITLE
S3 Storage ACL `public-read`

### DISCUSSION THREAD

**Discussion by ddcollins** [directus/directus#4681] [Feature Requests]
I've configured Directus storage to S3 for image uploads, with the plan of serving the images to our user base through our CDN and circumventing `/assets` and the API. However, on upload the ACL option is defaulting to private, which blocks us from accessing the images from S3 publicly, even though the S3 bucket is configured to be public within S3 itself.

In Directus 8 there was a way to define the `ACL` option: https://v8.docs.directus.io/extensions/storage-adapters.html#using-aws-s3 which would solve this issue, however I do not see a way to specify this as an env variable in Directus 9.

**Comment by rijkvanzanten** [comment]
We'll have to update the `@directus/drive-s3` package to introduce this configuration flag. Wanna open a PR for that?

**Reply by ddcollins** [reply]
> We'll have to update the `@directus/drive-s3` package to introduce this configuration flag. Wanna open a PR for that?

Got a working solution, but noticed that accessing the URL in S3 is triggering a download instead of rendering in the browser.

I believe this is because the type is defaulting to `'application/octet-stream'` because of this default:
https://github.com/directus/directus/blob/main/api/src/services/files.ts#L48

Is it possible to modify this default currently? And if not, thoughts on implementing a solution for it?

![Image 2021-03-26 at 11 57 57 AM](https://user-images.githubusercontent.com/38921476/112666827-9127b000-8e2a-11eb-8c54-83d84652932d.jpg)

**Reply by rijkvanzanten** [reply]
Hmm that default isn't send up to S3, so I don't think those two are related 🤔 It's only the default that's stored in directus_files.

The file stream is passed on straight from here: https://github.com/directus/directus/blob/main/api/src/services/files.ts#L52

Which is uploaded using busboy over here: https://github.com/directus/directus/blob/main/api/src/services/files.ts#L52



Do you have an example file where this is happening to test with?

**Reply by ddcollins** [reply]
> Hmm that default isn't send up to S3, so I don't think those two are related 🤔 It's only the default that's stored in directus_files.
> 
> The file stream is passed on straight from here: https://github.com/directus/directus/blob/main/api/src/services/files.ts#L52
> 
> Which is uploaded using busboy over here: https://github.com/directus/directus/blob/main/api/src/services/files.ts#L52
> 
> Do you have an example file where this is happening to test with?

You know what, digging further I think it's actually S3 defaulting to that content type. If I specify the `ContentType` value here [AmazonWebServicesS3Storage.ts#L253](https://github.com/directus/directus/blob/main/packages/drive-s3/src/AmazonWebServicesS3Storage.ts#L253) it works:
```
const params = {
	Key: location,
	Body: content,
	Bucket: this.$bucket,
	ACL: this.$acl,
	ContentType: 'image/png',
};
```

Inside the `put` method the content type needs to somehow get passed so we can send it up to S3.

**Reply by rijkvanzanten** [reply]
Ah ok! Quick question: does directus actually extract the correct mime type? If so, we could automatically send up the mimetype alongside the file itself to store it in S3 correctly (though I would've expected S3 to extract the mimetype from the stream automatically 🤔 )

**Reply by rijkvanzanten** [reply]
Also, just to rule out any other funkyness, does this happen with every file? Or does this problem only occur for a specific (set of) file(s) ?

**Reply by ddcollins** [reply]
> Ah ok! Quick question: does directus actually extract the correct mime type? If so, we could automatically send up the mimetype alongside the file itself to store it in S3 correctly (though I would've expected S3 to extract the mimetype from the stream automatically 🤔 )

Yes, this seems to be happening for all file types (just tested with a variety including PDF).

Directus looks to be extracting the correct mime type as the `type` property in the `directus_files` table is accurate `image/png`, `image/jpeg`, `application/pdf` etc.

However, [files.ts#L47](https://github.com/directus/directus/blob/main/api/src/services/files.ts#L47) the payload type here doesn't seem to be set... passing it along to `ContentType` on upload doesn't yield the same results as when I correctly hardcode `ContentType` to the corresponding content type of the file I'm uploading.

Unable to get my console logs spitting out in my terminal so can't be certain what the payload type is during execution.

**Reply by ddcollins** [reply]
@rijkvanzanten I was modifying the `ts` file but the `js` file in the dist folder was still outdated 🤦  All looks good now. Opened the PR.
https://github.com/directus/directus/pull/4694/files

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
